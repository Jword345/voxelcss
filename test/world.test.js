'use strict';

const Voxel = require('../lib/Voxel');
const Scene = require('../lib/Scene');
const World = require('../lib/World');

describe('World', () => {
    let label;
    let scene;
    let world;
    beforeEach(() => {
        label = 'Za Worudo';// Jo Jo reference
        scene = new Scene();
        world = new World(scene, label);
    });
    afterEach(() => {
        localStorage.setItem.mockClear();
        localStorage.getItem.mockClear();
    });
    it('can not be instantiated without Scene instance', () => {
        expect(() => {
            world = new World();
        }).toThrowErrorMatchingSnapshot();
    });
    it('can be instantiated without a label', () => {
        world = new World(scene);
        world.save();
        expect(localStorage.setItem).toHaveBeenLastCalledWith('savedWorld<*>', '[]');
    });
    it('can save and delete state', () => {
        const formName = `savedWorld<${label}>`;
        world.isSaved();
        expect(localStorage.getItem).toHaveBeenLastCalledWith(formName);
        world.save();
        expect(localStorage.setItem).toHaveBeenCalledTimes(1);
        expect(localStorage.setItem).toHaveBeenCalledWith(formName, '[]');
        world.deleteSave();
        expect(localStorage.setItem).toHaveBeenLastCalledWith(formName, '');
    });
    it('can export voxels', () => {
        world.add(new Voxel());
        expect(world.export()).toMatchSnapshot();
    });
    it('can import voxels', () => {
        world.add(new Voxel());
        const voxelString = world.export();
        world.import(voxelString);
        expect(world.export()).toEqual(voxelString);
        expect(() => {
            world.import(100);// not a string
        }).toThrowErrorMatchingSnapshot();
        expect(world.export()).toEqual(voxelString);
        expect(() => {
            world.import('');
        }).toThrowErrorMatchingSnapshot();
        expect(world.export()).not.toEqual(voxelString);// clear is called
        expect(world.export()).toMatchSnapshot();
    });
    it('can load from the browser', () => {
        const formName = `savedWorld<${label}>`;
        const voxel = new Voxel();
        world.add(voxel);
        expect(world.getVoxels()).toMatchSnapshot();
        world.save();
        world.remove(voxel);
        expect(world.getVoxels()).toEqual([]);
        world.load();
        expect(world.getVoxels()).toMatchSnapshot();
        expect(localStorage.getItem).toHaveBeenCalledTimes(1);
        localStorage.removeItem(formName);
        world.load();
        expect(world.getVoxels()).toMatchSnapshot();
    });
    it('can get scene and voxels', () => {
        expect(world.getScene()).toMatchSnapshot();
        expect(world.getVoxels()).toEqual([]);
        world.add(new Voxel());
        expect(world.getVoxels()).toMatchSnapshot();
    });
    it('can add and remove voxels', () => {
        const voxel = new Voxel();
        world.add(voxel);
        expect(world).toMatchSnapshot();
        const success = world.remove(voxel);
        expect(success).toBeTruthy();
        expect(world.export()).toEqual('[]');
        expect(world.remove()).toBeFalsy();
    });
});
