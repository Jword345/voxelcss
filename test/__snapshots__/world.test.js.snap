// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`World can add and remove voxels 1`] = `
World {
  "add": [Function],
  "deleteSave": [Function],
  "export": [Function],
  "getScene": [Function],
  "getVoxels": [Function],
  "import": [Function],
  "isSaved": [Function],
  "load": [Function],
  "remove": [Function],
  "save": [Function],
}
`;

exports[`World can export voxels 1`] = `"[{\\"position\\":{\\"x\\":0,\\"y\\":0,\\"z\\":0},\\"dimension\\":0,\\"mesh\\":\\"{\\\\\\"front\\\\\\":\\\\\\"image(data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7)\\\\\\",\\\\\\"back\\\\\\":\\\\\\"image(data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7)\\\\\\",\\\\\\"left\\\\\\":\\\\\\"image(data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7)\\\\\\",\\\\\\"right\\\\\\":\\\\\\"image(data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7)\\\\\\",\\\\\\"top\\\\\\":\\\\\\"image(data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7)\\\\\\",\\\\\\"bottom\\\\\\":\\\\\\"image(data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7)\\\\\\"}\\"}]"`;

exports[`World can get scene and voxels 1`] = `
Scene {
  "add": [Function],
  "addEventListener": [Function],
  "addLightSource": [Function],
  "attach": [Function],
  "bind": [Function],
  "canPan": [Function],
  "canRotate": [Function],
  "canZoom": [Function],
  "detach": [Function],
  "disablePan": [Function],
  "disableRotate": [Function],
  "disableZoom": [Function],
  "enablePan": [Function],
  "enableRotate": [Function],
  "enableZoom": [Function],
  "getElement": [Function],
  "getInteractionState": [Function],
  "getLightSources": [Function],
  "getPan": [Function],
  "getParentElement": [Function],
  "getRotation": [Function],
  "getRotationX": [Function],
  "getRotationY": [Function],
  "getRotationZ": [Function],
  "getVoxels": [Function],
  "getZoom": [Function],
  "isAttached": [Function],
  "pan": [Function],
  "panX": [Function],
  "panY": [Function],
  "panZ": [Function],
  "remove": [Function],
  "removeEventListener": [Function],
  "removeLightSource": [Function],
  "rotate": [Function],
  "rotateX": [Function],
  "rotateY": [Function],
  "rotateZ": [Function],
  "setPan": [Function],
  "setPanX": [Function],
  "setPanY": [Function],
  "setPanZ": [Function],
  "setRotation": [Function],
  "setRotationX": [Function],
  "setRotationY": [Function],
  "setRotationZ": [Function],
  "setZoom": [Function],
  "triggerEvent": [Function],
  "unbind": [Function],
  "zoom": [Function],
}
`;

exports[`World can get scene and voxels 2`] = `
Array [
  Voxel {
    "addEventListener": [Function],
    "addToScene": [Function],
    "animDown": [Function],
    "animUp": [Function],
    "clone": [Function],
    "getDimension": [Function],
    "getDomElement": [Function],
    "getMesh": [Function],
    "getPosition": [Function],
    "getPositionX": [Function],
    "getPositionY": [Function],
    "getPositionZ": [Function],
    "removeEventListener": [Function],
    "removeFromScene": [Function],
    "removeParentScene": [Function],
    "setDimension": [Function],
    "setMesh": [Function],
    "setParentScene": [Function],
    "setPosition": [Function],
    "setPositionX": [Function],
    "setPositionY": [Function],
    "setPositionZ": [Function],
    "translate": [Function],
    "translateX": [Function],
    "translateY": [Function],
    "translateZ": [Function],
    "triggerEvent": [Function],
    "updateLightSource": [Function],
  },
]
`;

exports[`World can load from the browser 1`] = `
Array [
  Voxel {
    "addEventListener": [Function],
    "addToScene": [Function],
    "animDown": [Function],
    "animUp": [Function],
    "clone": [Function],
    "getDimension": [Function],
    "getDomElement": [Function],
    "getMesh": [Function],
    "getPosition": [Function],
    "getPositionX": [Function],
    "getPositionY": [Function],
    "getPositionZ": [Function],
    "removeEventListener": [Function],
    "removeFromScene": [Function],
    "removeParentScene": [Function],
    "setDimension": [Function],
    "setMesh": [Function],
    "setParentScene": [Function],
    "setPosition": [Function],
    "setPositionX": [Function],
    "setPositionY": [Function],
    "setPositionZ": [Function],
    "translate": [Function],
    "translateX": [Function],
    "translateY": [Function],
    "translateZ": [Function],
    "triggerEvent": [Function],
    "updateLightSource": [Function],
  },
]
`;

exports[`World can load from the browser 2`] = `
Array [
  Voxel {
    "addEventListener": [Function],
    "addToScene": [Function],
    "animDown": [Function],
    "animUp": [Function],
    "clone": [Function],
    "getDimension": [Function],
    "getDomElement": [Function],
    "getMesh": [Function],
    "getPosition": [Function],
    "getPositionX": [Function],
    "getPositionY": [Function],
    "getPositionZ": [Function],
    "removeEventListener": [Function],
    "removeFromScene": [Function],
    "removeParentScene": [Function],
    "setDimension": [Function],
    "setMesh": [Function],
    "setParentScene": [Function],
    "setPosition": [Function],
    "setPositionX": [Function],
    "setPositionY": [Function],
    "setPositionZ": [Function],
    "translate": [Function],
    "translateX": [Function],
    "translateY": [Function],
    "translateZ": [Function],
    "triggerEvent": [Function],
    "updateLightSource": [Function],
  },
]
`;

exports[`World can load from the browser 3`] = `Array []`;

exports[`World can not be instantiated without Scene instance 1`] = `undefined`;
